### PR TITLE
fix paths w shortcuts when copying from pods

### PR DESCRIPTION
**Pull request by juanvallejo**
Addresses an issue where copying from a remote location containing path
shortcuts (podName:../../../tmp/foo) causes an index out of range panic.

**Release note**:
```release-note
The "kubectl cp" command now supports path shortcuts (../) in remote paths.
```

cc @soltysh

### CODE REVIEW

**Review by soltysh — APPROVED** [review]
/lgtm
/approve

You might need to cherry-pick that to 1.11.

**Comment by liggitt** [comment]
/hold 
Can you describe what this is doing? It's a little hard to follow and I want to make sure this isn't reintroducing the path traversal issues with cp fixed in 1.9

**Comment by juanvallejo** [comment]
@liggitt 
> Can you describe what this is doing? It's a little hard to follow and I want to make sure this isn't reintroducing the path traversal issues with cp fixed in 1.9

This addresses a panic that occurs when a user attempts to copy from a pod, specifying a relative path containing `../`:

```
$ kubectl cp mypod:../../../tmp/myfile ./
```

The panic was [happening here](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/cp.go#L404) due to the prefix being longer than the `header.Name` because of the path shortcuts "../../"

File: pkg/kubectl/cmd/cp.go [hunk 1]
```diff
@@ -299,6 +299,7 @@ func (o *CopyOptions) copyFromPod(src, dest fileSpec) error {
 // stripPathShortcuts removes any leading or trailing "../" from a given path
 func stripPathShortcuts(p string) string {
 	newPath := path.Clean(p)
+	newPath = strings.TrimLeft(p, "../")
```
Comment (liggitt) [inline pkg/kubectl/cmd/cp.go]: TrimLeft is not correct here. I think you meant TrimPrefix. TrimLeft removes all "/" and "." characters from the front of the string. It would turn ".../.../.../test" into "test"

File: pkg/kubectl/cmd/cp_test.go [hunk 2]
```diff
@@ -113,10 +113,14 @@ func TestGetPrefix(t *testing.T) {
 	}{
 		{
 			input:    "/foo/bar",
-			expected: "foo/bar",
```
Comment (liggitt) [inline pkg/kubectl/cmd/cp_test.go]: this is a pretty different result, and doesn't involve any backsteps in the path... why did this change?

File: pkg/kubectl/cmd/cp.go [hunk 3]
```diff
@@ -453,8 +454,24 @@ func untarAll(reader io.Reader, destFile, prefix string) error {
 }
 
 func getPrefix(file string) string {
+	segs := strings.Split(file, "/")
+	prefix := file
+
+	lastIdx := -1
+	for i := len(segs) - 1; i > 0; i-- {
+		if len(segs[i]) == 0 {
```
Comment (liggitt) [inline pkg/kubectl/cmd/cp.go]: can you describe the purpose of this function, and of this change?

**Comment by liggitt** [comment]
this should not be merged as-is. it looks like this was already picked to 1.11 and merged... we should probably revert it and wait until this is ready to reintroduce. the issue this is fixing was not a regression in 1.11, correct?

**Comment by juanvallejo** [comment]
> this should not be merged as-is. it looks like this was already picked to 1.11 and merged... we should probably revert it and wait until this is ready to reintroduce.

Okay

> the issue this is fixing was not a regression in 1.11, correct?

Not a regression - just a bug that had not been caught yet

**Comment by juanvallejo** [comment]
> thanks. can you open the revert against the 1.11 branch and tag sig-release and sig-cli on it?

https://github.com/kubernetes/kubernetes/pull/65336

File: pkg/kubectl/cmd/cp.go (hunk 3 above)
Comment (juanvallejo) [reply pkg/kubectl/cmd/cp.go]: it modified the remote path to just return the prefix without the filename, i.e. for a string "foo/bar/baz.exe" it returned "foo/bar/". I realize this changes existing behavior, particularly when attempting to copy a remote file under a different filename locally (copying remote file "/foo/bar/baz" to local file "buz" would now just create a directory "buz" and place file "baz" in it). Will revert this

File: pkg/kubectl/cmd/cp.go (hunk 1 above)
Comment (juanvallejo) [reply pkg/kubectl/cmd/cp.go]: Updated to use strings.TrimPrefix(). Since that method only trims the first instance, I kept the call to `path.Clean()` above (to first resolve cases where "../foo/../foo/bar/" is given), then remove any remaining instances of "../" that occur as a prefix

**Comment by juanvallejo** [comment]
@liggitt thanks, comments addressed

**Comment by juanvallejo** [comment]
@liggitt friendly ping

File: pkg/kubectl/cmd/cp_test.go [hunk 4]
```diff
@@ -128,6 +128,32 @@ func TestGetPrefix(t *testing.T) {
...
+		{
+			name:     "test single path shortcut prefix",
+			input:    "../foo/bar",
+			expected: "foo/bar",
+		},
+		{
+			name:     "test multiple path shortcuts",
+			input:    "../../foo/bar",
+			expected: "foo/bar",
+		},
```
Comment (liggitt) [inline pkg/kubectl/cmd/cp_test.go]: are the following inputs tested (both as unit tests and that they actually work when used in `cp`):
* absolute paths ("/foo/bar")
* parent paths with no named directory ("../../")
* parent paths with no named directory with no trailing slash ("../..")

Comment (juanvallejo) [reply pkg/kubectl/cmd/cp_test.go]: Thanks, added test in last commit

**Comment by juanvallejo** [comment]
@liggitt friendly ping

File: pkg/kubectl/cmd/cp.go [hunk 5]
```diff
@@ -299,6 +299,16 @@ func (o *CopyOptions) copyFromPod(src, dest fileSpec) error {
...
 	newPath := path.Clean(p)
+	trimmed := strings.TrimPrefix(newPath, "../")
+
+	for trimmed != newPath {
+		newPath = trimmed
+		trimmed = strings.TrimPrefix(newPath, "../")
+	}
+
+	// trim leftover ".."
+	newPath = strings.TrimPrefix(newPath, "..")
```
Comment (liggitt) [inline pkg/kubectl/cmd/cp.go]: won't this incorrectly remove ".." from "...foo"?

Comment (liggitt) [reply pkg/kubectl/cmd/cp.go]: add a testcase for a path containing three leading dots, since that has cropped up a few times as a mishandled case

File: test/e2e/kubectl/kubectl.go [hunk 6]
```diff
@@ -1046,6 +1048,46 @@ metadata:
...
+		AfterEach(func() {
+			cleanupKubectlInputs(podYaml, ns, busyboxPodSelector)
+		})
+
+		/*
+			Release : v1.12
+			Testname: Kubectl, copy
+			Description: When a Pod is running, copy a known file from it to a temporary local destination.
+		*/
+		framework.ConformanceIt("should copy a file from a running Pod ", func() {
```
Comment (liggitt) [inline test/e2e/kubectl/kubectl.go]: I'm not convinced this should be a conformance test. let's start with it as a normal e2e, and deal with promotion to conformance as an optional follow-up

File: test/e2e/kubectl/kubectl.go [hunk 7]
```diff
@@ -1046,6 +1048,46 @@ metadata:
 		})
 	})
 
+	framework.KubeDescribe("Kubectl copy", func() {
+		podYaml := substituteImageName(string(readTestFileOrDie("busybox-pod.yaml.in")))
```
Comment (liggitt) [inline test/e2e/kubectl/kubectl.go]: just name this file .yaml if it is a valid yaml file that can be created directly

**Comment by soltysh** [comment]
@juanvallejo I see the comments from Jordan are still not addressed,

**Comment by juanvallejo** [comment]
@soltysh @liggitt thanks, review comments addressed. PTAL

**Review by soltysh — APPROVED** [review]
/lgtm
/approve
/hold cancel
